### PR TITLE
add redirect to host rules for dns issue in kind

### PR DESCRIPTION
**Pull request by hedibouattour**
A dns issue showed in kind cluster, due to specific iptables rules added by docker. This commit addresses that by adding a feature allowing to redirect some traffic to host, to get it pass through iptables.
This is applied to coredns pods.
We use some new user config to specify rulesfor this feature. A tag "host" is added to cnat snat poliy feature and we use it for our vpptap0 to disable snat on traffic outgoing through it.

### CODE REVIEW

File: calico-vpp-agent/cni/cni_server.go [hunk 1]
```diff
@@ -442,6 +499,27 @@ forloop:
 	return nil
 }
 
+func (s *Server) createRedirectToHostRules() (uint32, error) {
+	var missNextIndex int32 = -1
+	index, err := s.vpp.AddClassifyTable(&types.ClassifyTable{
+		Mask:           types.DstThreeTupleMask,
+		NextTableIndex: types.InvalidID,
+		MaxNumEntries:  200000,
```
Comment (sknat) [inline calico-vpp-agent/cni/cni_server.go:507]: `200k` is probably a bit much here 😄 
we could use something like `2 * len(config.redirects)`

File: calico-vpp-agent/cni/cni_server.go [hunk 2]
```diff
@@ -442,6 +499,27 @@ forloop:
 	return nil
 }
 
+func (s *Server) createRedirectToHostRules() (uint32, error) {
+	var missNextIndex int32 = -1
```
Comment (sknat) [inline calico-vpp-agent/cni/cni_server.go:503]: Not sure why we define this ?

File: calico-vpp-agent/cni/cni_server.go [hunk 3]
```diff
@@ -315,9 +324,57 @@ func (s *Server) rescanState() {
...
+			if err != nil {
+				s.log.Error(err)
+			}
+		}
+	}
+}
+
+func (s *Server) PodNeedsRedirectToHost(podSpec *storage.LocalPodSpec) bool {
+	return strings.HasPrefix(podSpec.WorkloadID, "kube-system/coredns")
+}
```
Comment (sknat) [inline calico-vpp-agent/cni/cni_server.go:338]: No need to restrict this just to the coredns, arguably all rules should apply to all pods

File: config/config.go [hunk 4]
```diff
@@ -252,6 +258,17 @@ func (u *UplinkInterfaceSpec) String() string {
...
 }
 
+type RedirectToHostRulesConfigType struct {
+	Port uint16 `json:"port,omitempty"`
+	Ip   string `json:"ip,omitempty"`
+	/* "tcp", "udp",... */
+	Proto types.IPProto `json:"proto,omitempty"`
+}
+
+type SnatExclusionPrefixesConfigType struct {
```
Comment (sknat) [inline config/config.go:268]: Could we expose a list of exclusion prefixes ?

File: config/config.go [hunk 5]
```diff
@@ -252,6 +258,17 @@ func (u *UplinkInterfaceSpec) String() string {
 	return string(b)
 }
 
+type RedirectToHostRulesConfigType struct {
```
Comment (sknat) [inline config/config.go:261]: Same for rules, a list could be good

File: vpplink/classify.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	response, err := client.ClassifyAddDelTable(v.GetContext(), &classify.ClassifyAddDelTable{
+		IsAdd:             isAdd,
+		DelChain:          delChain,
+		TableIndex:        tableIndex,
+		Nbuckets:          nBuckets,
+		MemorySize:        memorySize,
+		SkipNVectors:      table.SkipNVectors,
+		MatchNVectors:     matchNVectors,
+		NextTableIndex:    table.NextTableIndex,
+		MissNextIndex:     table.MissNextIndex, //ip4-input
```
Comment (sknat) [inline vpplink/classify.go:79]: A rogue comment that slipped through 😄

File: yaml/overlays/dev/calico-vpp-dev-configmap.yaml [hunk 7]
```diff
@@ -4,6 +4,8 @@ metadata:
   name: calico-vpp-config
   namespace: calico-vpp-dataplane
 data:
+  CALICOVPP_SNAT_EXCLUSION_PREFIXES: ${CALICOVPP_SNAT_EXCLUSION_PREFIXES}
+  CALICOVPP_REDIRECT_TO_HOST_RULES: ${CALICOVPP_REDIRECT_TO_HOST_RULES}
```
Comment (sknat) [inline yaml/overlays/dev/calico-vpp-dev-configmap.yaml:8]: Maybe we could not put these inside the `CALICOVPP_INITIAL_CONFIG` envvar no ?

**Review by sknat — COMMENTED** [review]
Many thanks for this PR @hedibouattour,
a few comments inline ! 😄

File: vpplink/classify.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+import (
+	"fmt"
+
+	"github.com/projectcalico/vpp-dataplane/v3/vpplink/generated/bindings/classify"
+	"github.com/projectcalico/vpp-dataplane/v3/vpplink/generated/bindings/interface_types"
+	"github.com/projectcalico/vpp-dataplane/v3/vpplink/types"
+)
+
+type lo uint8
```
Comment (sknat) [inline vpplink/classify.go:26]: This seems unsued

File: calico-vpp-agent/cni/cni_server.go [hunk 9]
```diff
@@ -315,9 +324,57 @@ func (s *Server) rescanState() {
...
+			} else {
+				s.log.Infof("set input acl table %d for interface %d successfully", s.RedirectToHostClassifyTableIndex, swIfIndex)
+				done <- true
+				return
+			}
+			times += 1
+			if times >= timeout {
+				done <- false
+				return
+			}
```
[the remark this answers is not in the document]
Comment (hedibouattour) [reply calico-vpp-agent/cni/cni_server.go:368]: indeed no need for it

File: calico-vpp-agent/cni/cni_server.go [hunk 10]
```diff
@@ -453,6 +531,10 @@ func (s *Server) ServeCNI(t *tomb.Tomb) error {
 		return errors.Wrapf(err, "failed to listen on %s", config.CNIServerSocket)
 	}
 
+	s.RedirectToHostClassifyTableIndex, err = s.createRedirectToHostRules()
```
[the remark this answers is not in the document]
Comment (hedibouattour) [reply calico-vpp-agent/cni/cni_server.go:534]: This implem does not break anything, it creates a rule that would match a protocol = 0 and a port =0  so it matches nothing. But yes it's better to have it only when defined.

**Pull request merged**